File: python/01-fundamentals/04-strings/solution.py
# solution.py - Answers and explanations for string exercises

# TODO 1: Create a variable 'greeting' and assign it "Hello, Python!"
greeting = "Hello, Python!"

# TODO 2: Calculate the length of 'greeting' using the len() function.
length = len(greeting)

# TODO 3: Extract the word "Python" from the 'greeting' string using slicing.
# "Hello, " is 7 chars (0-6). "Python" starts at index 7 and ends at 12.
# So slice is [7:13] (stop index 13 is not included).
python_word = greeting[7:13]

# TODO 4: Create a new string 'shout' that is the uppercase version of 'greeting'.
shout = greeting.upper()

# TODO 5: Replace "Hello" with "Welcome" in the 'greeting' string.
welcome = greeting.replace("Hello", "Welcome")

# TODO 6: Use an f-string to create a message: "The length of '{greeting}' is {length}."
message = f"The length of '{greeting}' is {length}."

# CHALLENGE: Reverse the 'greeting' string using slicing.
reversed_greeting = greeting[::-1]

# --- Verification ---
print("--- Verification ---")
print(f"Greeting: {greeting}")
print(f"Length: {length}")
print(f"Python word: {python_word}")
print(f"Shout: {shout}")
print(f"Welcome: {welcome}")
print(f"Message: {message}")
print(f"Reversed: {reversed_greeting}")

print("\n--- Why it works ---")
print("1. slicing format is [start:stop:step].")
print("2. string methods (upper, replace) return a NEW string, they don't change the original.")
print("3. negative index starts from the end (-1 is the last char).")
print("4. [::-1] reversed the string because we step backward through the entire sequence.")

File: javascript/01-fundamentals/04-strings/solution.js
// solution.js - String answers

// TODO 1
const product = "Laptop";
const price = 999;
console.log(`The ${product} costs $${price}.`);

// TODO 2
const poem = `
Roses are red,
Violets are blue,
JavaScript is cool,
And so are you.
`;

// TODO 3
const message = "  HELLO WORLD  ";
const cleaned = message.trim().toLowerCase();

// TODO 4
const sentence = "I love coding";
const hasLove = sentence.includes("love");

// TODO 5
const csv = "apple,orange,banana";
const fruits = csv.split(",");

// TODO 6
const fullText = "Learning JS is fun";
const onlyJS = fullText.slice(9, 11);

// CHALLENGE ANSWER
function capitalize(word) {
    if (!word) return "";
    return word[0].toUpperCase() + word.slice(1).toLowerCase();
}

// --- Verification ---
console.log("--- Verification ---");
console.log(`Cleaned: '${cleaned}'`);
console.log(`Has 'love'? ${hasLove}`);
console.log(`Fruits:`, fruits);
console.log(`Slice result: ${onlyJS}`);
console.log(`Capitalize('python'): ${capitalize('python')}`);

console.log("\n--- Why it works ---");
console.log("1. Template Literals: Use backticks `` and ${} to embed values directly, avoiding messy + concatenation.");
console.log("2. Chaining: Methods return new strings, allowing you to chain like .trim().toLowerCase().");
console.log("3. Immutability: String methods NEVER change the original string. They return a copy.");
console.log("4. Array-like: You can access characters by index (word[0]) just like in an array.");
